File: leetcode/go/123.best-time-to-buy-and-sell-stock-iii.go
func maxProfit(prices []int) int {
    hold1, sell1, hold2, sell2 := math.MinInt32, 0, math.MinInt32, 0

    for _, price := range prices {
        sell2 = max(sell2, hold2 + price)
        hold2 = max(hold2, sell1 - price)
        sell1 = max(sell1, hold1 + price)
        hold1 = max(hold1, -price)
        // fmt.Println(price, sell2, hold2, sell1, hold1)
    }

    return sell2
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: leetcode/rust-leetcode/src/123.best-time-to-buy-and-sell-stock-iii.rs
use std::cmp::max;
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let (mut hold1, mut sell1, mut hold2, mut sell2) = (i32::MIN, 0, i32::MIN, 0);

        for price in prices {
            // println!("{price} {sell2} {hold2} {sell1} {hold1}");
            sell2 = max(sell2, hold2 + price);
            hold2 = max(hold2, sell1 - price);
            sell1 = max(sell1, hold1 + price);
            hold1 = max(hold1, -price);
            // println!("{price} {sell2} {hold2} {sell1} {hold1}");
        }

        sell2
    }
}

